### ISSUE TITLE
Can client upload data when was not initialized from server?

### ISSUE THREAD

**Issue by gentledepp** [Mimetis/Dotmim.Sync#40]
Hi!

I recently stumbled across a shortcoming of Microsoft Sync Framework again and wanted to know if this was fixed in doxmim sync.

With Sync Framework, you had to 
* locally provision the sync tables (in the client sqlite database)
* start an initial sync with the server
=> only then it was possible, to alter data in the client sqlite database and push those changes back to the server.

If you added data before the initial sync with the server, you'd get a "serverBlob is null" exception.

But sometimes, a user has some app, which is not yet connected to a server and (s)he adds data anyways. When online again, (s)he naturally wants to push those changes to the server nontheless.

Is this possible with dotmimsync, or do you still have to run an initial sync before editing data?

**Comment by Mimetis** [comment]
Good question !!   

### TL;DR 

Short answer is **YES**

### Example

Basically, the starter solution is this one:
You have both an **already existing SQLite** client database and an **already existing SQL Server** master database.  
I assume you're aware that the **schema** should be **exactly the same** beetween the server and the client, since it's NOT dedicated to the `Dotmim.Sync` fx (to create the client schema)   

So to start, imagine we have those 2 databases:

Here is the **SQL Server** database, acting as the `master`:
```sql
if (not exists (select * from sys.tables where name = 'ServiceTickets'))
begin
   CREATE TABLE [ServiceTickets](
     [ServiceTicketID] [uniqueidentifier] NOT NULL,
     [Title] [nvarchar](max) NOT NULL,
     [StatusValue] [int] NOT NULL,
     [Opened] [datetime] NULL,
   CONSTRAINT [PK_ServiceTickets] PRIMARY KEY CLUSTERED ( [ServiceTicketID] ASC ));
end
```
And then the **sqlite** table as the `client`:
```sql
CREATE TABLE IF NOT EXISTS [ServiceTickets](
  [ServiceTicketID] blob NOT NULL,
  [Title] text NOT NULL,
  [StatusValue] integer NOT NULL,
  [Opened] datetime NULL,
PRIMARY KEY ( [ServiceTicketID] ASC ))
```
Then we can add 1 row in each database, **before** engaging any sync (even any sync configuration) with this kind of method:

``` csharp
public void AddingDatas(DbConnection connection)
{

    var command = connection.CreateCommand();
    command.CommandText = $@"INSERT INTO [ServiceTickets] ([ServiceTicketID], [Title], [StatusValue], [Opened]) 
                                VALUES (@ServiceTicketID, @Title, @StatusValue, @Opened)";

    DbParameter parameter = null;
    parameter = command.CreateParameter();
    parameter.DbType = DbType.Guid;
    parameter.ParameterName = "@ServiceTicketID";
    parameter.Value = Guid.NewGuid();
    command.Parameters.Add(parameter);

    parameter = command.CreateParameter();
    parameter.DbType = DbType.String;
    parameter.ParameterName = "@Title";
    parameter.Value = $"Title - {Guid.NewGuid().ToString()}";
    command.Parameters.Add(parameter);

    parameter = command.CreateParameter();
    parameter.DbType = DbType.Int32;
    parameter.ParameterName = "@StatusValue";
    parameter.Value = new Random().Next(0, 10);
    command.Parameters.Add(parameter);

    parameter = command.CreateParameter();
    parameter.DbType = DbType.DateTime;
    parameter.ParameterName = "@Opened";
    parameter.Value = DateTime.Now;
    command.Parameters.Add(parameter);

    try
    {
        connection.Open();
        command.ExecuteNonQuery();
        connection.Close();
    }
    catch (Exception ex)
    {
        Debug.WriteLine(ex.Message);
        throw;
    }
    finally
    {
        if (connection.State != ConnectionState.Closed)
            connection.Close();
    }
}
```

Ok, right now, I have **One line on the server side** and **One line on the client side**.  
Be aware, we are at a starting point where **NO** Sync architecture is involved (no metadatas, no tracking tables and so on...)

Then, we can try a new sync processus:

```csharp
SqlSyncProvider serverProvider;
SqliteSyncProvider clientProvider;

serverProvider = new SqlSyncProvider(serverConnectionString);
clientProvider = new SqliteSyncProvider(clientSqliteFilePath);

var simpleConfiguration = new SyncConfiguration(new string[] { "ServiceTickets" });

agent = new SyncAgent(clientProvider, serverProvider, simpleConfiguration);

serverProvider.SyncProgress += (s, e) => Debug.WriteLine($"[Server]:{e.Message} {e.PropertiesMessage}");
clientProvider.SyncProgress += (s, e) => Debug.WriteLine($"[Client]:{e.Message} {e.PropertiesMessage}");

var session = await agent.SynchronizeAsync();
```

Here is an extract output debug I have:
```
[Client]:Changes selected TableName: ServiceTickets Deletes: 0 Inserts: 1 Updates: 0 TotalChanges: 1
[Server]:Applying changes TableName: ServiceTickets State: Added
[Server]:Changes applied TableName: ServiceTickets State: Added Applied: 1 Failed: 0
[Server]:Selecting changes TableName: ServiceTickets
[Server]:Changes selected TableName: ServiceTickets Deletes: 0 Inserts: 1 Updates: 0 TotalChanges: 1
[Client]:Applying changes TableName: ServiceTickets State: Added
[Client]:Changes applied TableName: ServiceTickets State: Added Applied: 1 Failed: 0
```
I made this long answer to be able to create a new chapter in the doc [https://mimetis.github.io/Dotmim.Sync/](https://mimetis.github.io/Dotmim.Sync/), so if something is not clear, do not hesitate to ping me :)

Sebastien

**Comment by gentledepp** [comment]
Tha'ts awesome! I did go through the documentation, though ;-)
However, this question did not cross my mind back then and I wanted to be absolutely sure this works.
Thank you so much for the clarification!

**Comment by Mimetis** [comment]
Ok, so I will close this issue, feel free to re open it, if you have more questions 

Seb
